src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs

```diff
@@ -0,0 +1,321 @@
+
+        internal static DiagnosticDescriptor StaticAbstractIsPreviewFeatureRule = DiagnosticDescriptorHelper.Create(RuleId,
+                                                                                                                    s_localizableTitle,
+                                                                                                                    s_staticAbstractMessage,
+                                                                                                                    DiagnosticCategory.Usage,
+                                                                                                                    RuleLevel.BuildError,
+                                                                                                                    s_localizableDescription,
+                                                                                                                    isPortedFxCopRule: false,
+                                                                                                                    isDataflowRule: false);
+        internal static DiagnosticDescriptor ImplementsEmptyPreviewInterfaceRule = DiagnosticDescriptorHelper.Create(RuleId,
+                                                                                                                    s_localizableTitle,
+                                                                                                                    s_implementsEmptyPreviewInterface,
+                                                                                                                    DiagnosticCategory.Usage,
+                                                                                                                    RuleLevel.BuildError,
+                                                                                                                    s_localizableDescription,
+                                                                                                                    isPortedFxCopRule: false,
+                                                                                                                    isDataflowRule: false);
+
+        private const string RequiresPreviewFeaturesAttribute = nameof(RequiresPreviewFeaturesAttribute);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(GeneralPreviewFeatureAttributeRule, StaticAbstractIsPreviewFeatureRule, ImplementsEmptyPreviewInterfaceRule);
+
+        private enum PreviewFeatureUsageType
+        {
+            General = 0,
+            StaticAbstract,
+            EmptyInterface
+        }
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze);
```

Comment:
Analyzing generated code without reporting the diagnostics doesn't make sense
```suggestion
            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
```

Reply:
@buyaa-n I believe it does. We want to collect data about generated code, but only report for non-generated code.

Reply:
Probably there is a case where it makes sense, especially when the diagnostic is not important, for this analyzer i believe it is important to report diagnostics, but i will let @pgovind decide

Reply:
@buyaa-n I thought it was agreed on that only security analyzers should report for auto-generated code?

Reply:
Oh yeah? Did not know what, do you have a reference to that agreement? I would like to see that if there is a reference. 

By accounting, this analyzer is error level and it will be quite expensive i would expect it to have either `GeneratedCodeAnalysisFlags.None` or if we need to analyze generated code it should report the diagnostics too, but i will let @pgovind  and repo owners decide CC @jmarolf @sharwell

Reply:
We should cover generated code as well.  Any use of a preview feature when not in preview, whether that use comes from handwritten or generated code, should be flagged.

Reply:
> do you have a reference to that agreement?

I don't recall where the comment was. But since this is the "first" analyzer with error severity, it may be an exception that it reports on generated code as well. cc @mavasani

Reply:
> But since this is the "first" analyzer with error severity, it may be an exception that it reports on generated code as well

That sounds a valid reason for this code to report diagnostics in generated code.